### PR TITLE
Feat/add linear

### PR DESCRIPTION
**Pull request by L1aoXingyu**
add linear1D, MLP and LayerNorm

### CODE REVIEW

File: libai/layers/linear.py [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+        out_features,
+        bias=True,
+        parallel="data",
+        activation=None,
+        bias_gelu_fusion=False,
+        output_dropout_prob=0.0,
+        bias_dropout_fusion=False,
+        init_method=nn.init.xavier_normal_,
+        *,
+        layer_idx=0,  # Enforce layer_idx passed with keyword
```
Comment (L1aoXingyu) [inline libai/layers/linear.py:63]: 这里主要考虑到如果用户不使用流水并行，那么这个参数可以先不暴露给用户
当用户需要用这个参数的话，要求他一定按照 `layer_idx=i` 这种方式显示传参，避免参数误传的问题

**Comment by L1aoXingyu** [comment]
将 linear, mlp 和 layernorm 都放在不同的文件夹主要考虑后面有可能对 linear2d, mlp2d 和 layernorm2d 的扩展

File: libai/layers/linear.py (hunk 1 above)
Comment (dangkai4u) [reply libai/layers/linear.py:63]: 1. 在attention模块中，megatron使用layer_idx对不同层的计算结果进行了放缩，也不清楚有没有用。不过，把layer_idx参数放后面，同时加上默认值0，感觉挺好的，可以让模块更通用，在单机单卡的情况下避免用户改动太多的代码。
2. 是不是把RowParallelLinear和ColumnParallelLinear保留上比较好，这两个模块中加fusion。在linear1D模块中，只实现linear操作，不添加激活函数和dropout.

Comment (L1aoXingyu) [reply libai/layers/linear.py:63]: rowparallel 和 columnParallel 后面都可以接 dropout fusion，他们也都可以在后面接 gelu fusion，只是 megatron 里面 col -> gelu fusion -> row -> dropout fusion，如果是 col -> gelu -> dropout 这种 以及 row -> gelu -> dropout 不是还要再写一个。

如果 columnParallel 和 rowParallel 只考虑 fusion 的话，他们俩的代码基本是一样的，除了中间的 linear1D 不一样。

File: libai/layers/mlp.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        *,
+        layer_idx=0,
+    ):
+        super().__init__()
+
+        self.dense_h_to_4h = Linear1D(
+            hidden_size,
+            ffn_hidden_size,
+            True,
+            "col",
```
Comment (CPFLAME) [inline libai/layers/mlp.py:54]: 这里写成 `bias=True` 和 `parallel="col"` 比较清晰一点,    在看这里的时候 我往上翻 看了函数的定义才知道True是啥

File: libai/layers/linear.py (hunk 1 above)
Comment (dangkai4u) [reply libai/layers/linear.py:63]: 分开写的好处：
1. fusion属于定制化操作，定制化目的是效率，只在特殊的情形下使用，减少不必要的操作或运算。linear1d的目的是通用，所以做到原子性即可，对标nn.Linear模块，增加三种分布式情形，不用增加太多的其他操作。
2. linear1d包含尽可能多的功能，用户用起来其实也很麻烦。tensorflow 2.0中也有linear+activation的算子，但使用的情形很少，这些完全可以让用户写。
3. 用户在熟悉分布式的情况下，可以选择正确的定制module。当不熟悉分布式时，采用linear1d，我们帮他们完成底层的boxing。这应该是我们想要的结果。既有效率，也有通用性。

Comment (L1aoXingyu) [reply libai/layers/linear.py:63]: 你可以基于这个分支开一个新的分支，然后把你想做的形式大概写出来，提一个pr，这样能更好的get到你的意思，光看文字我不知道该怎么操作

File: libai/layers/mlp.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        layer_idx: A layer_idx sign which determines the placement. It will be used in pipeline parallelism. Defaults to 0.
+    """
+
+    def __init__(
+        self,
+        hidden_size,
+        ffn_hidden_size,
+        output_dropout_prob=0.0,
+        init_method=nn.init.xavier_normal_,
+        output_layer_init_method=nn.init.xavier_normal_,
```
Comment (dangkai4u) [inline libai/layers/mlp.py:44]: output_layer_init_method默认值为None，然后在init函数中判断，如果为None，设为init_method。这样的话，如果使用相同的初始化函数，用户只需要指定一个。

File: libai/layers/linear.py (hunk 1 above)
Comment (dangkai4u) [reply libai/layers/linear.py:63]: 看了一下megatron和夸父，都采用skip_bias_add。要不咱们也添加一个skip_bias_add参数，然后把那两个fusion算子放到外面调用。linear模块不包含dropout和激活函数，这样linear就简洁，易用了。

Comment (L1aoXingyu) [reply libai/layers/linear.py:63]: 其实加激活函数和 dropout 都是为了 fuse bias_add，否则分开是很容易的，现在分开不容易就是因为 fuse 算子的问题

加在一起的好处是：用户可以简单当成 linear，不传其他参数就好了，也可以当成当成一个 fuse gelu/dropout 的 linear
坏处就是：linear 代码会变复杂

如果把 fusion 拆出去的好处就是代码会变得比较简洁，
坏处就是外面用的 fusion 的地方就得写重复写相同的代码，而且基本用的 linear 的地方几乎都有 fusion，不管是 activation 或者是 dropout，还有就是需要 reture bias，我觉得这种方式很奇怪

所以两种方式我觉得都不好，但是目前貌似也没有更好的方案，我们可以二选一，目前我对两种方案其实也没有明确的倾向性

Comment (dangkai4u) [reply libai/layers/linear.py:63]: 还是选简洁的吧。这两个fusion算子使用情况有限，主要是在MLP模块中使用，其他地方通常是普通的linear模块。用户的需求无法枚举，且除了attention和mlp模块，其他地方性能影响比较小，所以交给用户也没什么问题。

Comment (CPFLAME) [reply libai/layers/linear.py:63]: 我添加一点额外的信息:
额外的好处: 我记得当时glm项目了里面 用了fused kernel速度的提升还是比较明显的. 如果这么写, 用户想提速可以直接指认参数就可以了
额外的坏处: 我去问了一下晓雨,  fused kernel在转成onnx的时候是会报错的, 后期如果我们想提供一键转码onnx的脚本, 这个是需要注意的一个点: "训练的时候采用 fused kernel.  转onnx的时候把fused kennel得分开"

Comment (L1aoXingyu) [reply libai/layers/linear.py:63]: @Ldpe2G 德澎对 linear 这边还有额外意见吗，没有的话就按单独的 linear 实现来做了

Comment (Ldpe2G) [reply libai/layers/linear.py]: 暂时没有了

Comment (dangkai4u) [reply libai/layers/linear.py]: 建议设一个linear类，和linear1d内部实现一样。这个做法的好处在于，用户使用起来没有障碍。短期内，模型并行方式以1d为主。虽然之后我们会支持linear2d等模块，但这些模块的优势尚不清楚，如果有优势，到时候再推广。在那时，linear默认为1d并行，用户也可以通过linear1d或linear2d等显式地指明是哪种并行。

File: libai/layers/__init__.py [hunk 4]
```diff
@@ -14,3 +14,7 @@
 # limitations under the License.
 
 from .embedding import *
```
Comment (dangkai4u) [inline libai/layers/__init__.py:16]: embedding的import改一下写法吧，不用import *这种写法，把那几个模块导出来，然后在init文件里定义all

**Review by Ldpe2G — APPROVED** [review]
